aiosonic/connection.py

```diff
@@ -125,21 +134,43 @@ async def __aenter__(self):
 
     async def __aexit__(self, exc_type: None, exc: None, tb: None) -> None:
         """Release connection."""
-        if self.keep and not exc:
-            self.key = self.temp_key
-        else:
-            self.key = None
-            self.h2conn = None
-            if self.writer and not self.blocked:
+        try:
+            if self.keep and not exc:
+                self.key = self.temp_key
+            else:
+                self.key = None
+                self.h2conn = None
+
+            if (not self.keep or type(self.connector.pool) == CyclicQueuePool) and self.writer and not self.blocked:
                 self.close()
 
-        if not self.blocked:
-            await self.release()
-            if self.h2handler:  # pragma: no cover
-                self.h2handler.cleanup()
+            if not self.blocked and not self.cycled:
+                await self.release()
+                if self.h2handler:  # pragma: no cover
+                    self.h2handler.cleanup()
+
+            else:
+                while self.blocked:
```

Comment:
This `__aexit__` step is executed in a separated asyncio task? this doesn't block?

The CI (gh actions) with linux is blocked I guess it is for this :open_mouth:

Maybe it worth to create a separate task with `asyncio.create_task` that does this active wait (or to have a "cleaner" task for it)

Reply:
It does not block because of the nature of asyncio.sleep, which yields to other coroutines.

Sorry to hear about the CI issues.

We really can't have a task easily there I think because we need to catch CancelledError which is raise by wait_for in _do_request. I'll see if I can do it in a more elegant manner. Thanks.